inject/src/main/java/io/micronaut/context/AnnotationReflectionUtils.java

```diff
@@ -0,0 +1,242 @@
+            if (annotatedActualTypeArguments.length != typeParameters.length) {
+                throw new IllegalArgumentException("Annotated parameters must have the same number of type parameters");
+            }
+            List<Argument<?>> list = new ArrayList<>();
+            for (int i = 0; i < annotatedActualTypeArguments.length; i++) {
+                TypeVariable<? extends Class<?>> typeVariable = typeParameters[i];
+                Class<?> variableBaseType = null;
+                AnnotatedType resolvedVariable = resolvedVariables.get(typeVariable.getName());
+                if (resolvedVariable == null) {
+                    resolvedVariable = annotatedActualTypeArguments[i];
+                    TypeVariable<? extends Class<?>> typeParameter = typeParameters[i];
+                    variableBaseType = getClassFromType(typeParameter.getBounds()[0]);
+                }
+                if (resolvedVariable instanceof TypeVariable<?>) {
+                    // Found unresolved type variable
+                    return null;
+                }
+                Argument<?> argument = argumentOf(resolvedVariable, variableBaseType, resolvedVariables);
+                if (argument == null) {
+                    return null;
+                }
+                list.add(Argument.ofTypeVariable(
+                    argument.getType(),
+                    typeVariable.getName(),
+                    argument.getAnnotationMetadata(),
+                    argument.getTypeParameters()
+                ));
+            }
+            return Argument.of(
+                resolvedType,
+                annotationMetadataOf(type),
+                list.toArray(new Argument<?>[0])
+            );
+        }
+        return Argument.of(resolvedType, annotationMetadataOf(type));
+    }
+
+    private static AnnotationMetadata annotationMetadataOf(AnnotatedElement annotatedElement) {
+        Annotation[] annotations = annotatedElement.getAnnotations();
+        if (annotations.length == 0) {
+            return AnnotationMetadata.EMPTY_METADATA;
+        }
+        MutableAnnotationMetadata mutableAnnotationMetadata = new MutableAnnotationMetadata();
+        for (Annotation annotation : annotations) {
+            Map<CharSequence, Object> values = new LinkedHashMap<>();
+            Class<? extends Annotation> annotationType = annotation.annotationType();
+            Method[] methods = annotationType.getMethods();
+            for (Method method : methods) {
+                if (!method.getDeclaringClass().equals(annotationType)) {
+                    continue;
+                }
+                try {
+                    Object value = method.invoke(annotation);
+                    if (value != null) {
+                        values.put(method.getName(), value);
+                    }
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
```

Comment:
This should use `ReflectionUtils.invokeMethod`

Reply:
Changed